crunch-core/src/main/java/org/apache/crunch/io/SourceTargetHelper.java

```diff
@@ -41,17 +41,23 @@ public static long getPathSize(FileSystem fs, Path path) throws IOException {
     }
     long size = 0;
     for (FileStatus status : stati) {
-      if (status.isDir()) {
-        for (FileStatus st : fs.listStatus(status.getPath())) {
-          size += getPathSize(fs, st.getPath());
-        }
-      } else {
-        size += status.getLen();
-      }
+      size += getPathSize(fs, status);
     }
     return size;
   }
-  
+
+  private static long getPathSize(final FileSystem fs, final FileStatus status) throws IOException {
```

Comment:
This is still doing a recursive treewalk. IF you do `FileSystem.listFiles(path, true)` you get a deep listing of files only from the store. For an object store with an optimised implementation (as s3a does), only one HTTP request is made per 1000 objects, irrespective of the depth of the tree. For HDFS &c there's still efficiencies, especially when you have a directory with many millions of files in: in listStatus() all the results have to get serialized and marshalled over as one, rather than paged over to the client

Reply:
Thanks for the feedback @steveloughran. I have it updated locally to use `listFiles` as you suggest and it is working well, I just need to update the tests now before commiting.